---
title: "Sample `r params$sample` PreProcessing"
output:
  html_document:
    toc: true
    toc_float:
      collapsed: false
date: "`r Sys.Date()`"
params:
  workdir: seurat
  sample: sample_name
  celltype: NA
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message=FALSE, warning = FALSE, results='asis')
options(knitr.graphics.error = FALSE)
```

```{r Packages, message=FALSE}
library(knitr)
library(stringr)
library(flextable)
library(kableExtra)
# library(scater)
```

```{r Input, include=FALSE}
workdir <- params$workdir
sample <- params$sample
celltype <- params$celltype
hashtag <- FALSE
demux <- FALSE
```

Working Directory: `r workdir`

Sample Name: `r sample`


```{r Change Working Directory, include=FALSE}
knitr::opts_knit$set(root.dir = workdir)
```


## UMAP Plot and Clusters {.tabset}

### GEX-Based
```{r}
include_graphics(file.path(workdir, 'UMAP_RNA.png'))

include_graphics(file.path(workdir, 'UMAP_RNA_Split_Raster.png'))
```

### ADT-Based
```{r}
include_graphics(file.path(workdir, 'UMAP_ADT.png'))

include_graphics(file.path(workdir, 'UMAP_ADT_Split_Raster.png'))
```

<!-- ### Weighted Nearest Neighbor -->
<!-- ```{r} -->
<!-- include_graphics(file.path(workdir, 'UMAP_MultiModal_WNN.png')) -->

<!-- include_graphics(file.path(workdir, 'UMAP_MultiModal_WNN_Split_Raster.png')) -->
<!-- ``` -->

<!-- #### Weighted Nearest Neighbor Cell Modality Weights per Cluster -->
<!-- ```{r} -->
<!-- include_graphics(file.path(workdir, 'MultiModal_WNN_Weights.png')) -->
<!-- ``` -->

<!-- ### Weighted Nearest Neighbor Clusters with RNA and ADT UMAP Coordinates -->
<!-- ```{r} -->
<!-- include_graphics(file.path(workdir, 'UMAP_MultiModal_RNA_ADT.png')) -->
<!-- ``` -->

```{r Demuxlet Counts, results='asis', eval=file.exists('UMAP_RNA_Demux_Split_Raster.png')}
cat("## Demuxlet Results")
cat("\n\n")
##umap plot of group and split by genotype
##and a table outlining the cell numbers in each

include_graphics(file.path(workdir, 'UMAP_RNA_Demux_Split_Raster.png'))
```


## Gene Expression Cluster-based Differentially Expressed Genes {.tabset}
```{r, results='asis'}
filenames <- Sys.glob(file.path(workdir, 'DEG/*txt'))
clusters <- list()
for (filename in str_sort(filenames, numeric=TRUE)) {
  file_split <- strsplit(basename(filename), '_')[[1]]
  cluster <- file_split[2:(length(file_split)-1)]
  Markers <- read.table(filename, header=TRUE)
  clusters[[cluster]] <- Markers
}

degTable <- function(x){
  cat("### ", x)
  cat(knitr::knit_print(DT::datatable(clusters[[x]], height='100%', width='100%')))
}

output <- lapply(names(clusters)[1:(length(clusters)-1)], degTable)

cat("### ", names(clusters)[[length(clusters)]])
knitr::knit_print(DT::datatable(clusters[[length(clusters)]], height='100%', width='100%'))

```

```{r Cell Type Results, results='asis', out.height="80%", eval=celltype != 'NA'}
cat("## Gene Expression Based Cell Type Prediction {.tabset}")
cat("\n\n")

cat("### Level 1")
cat("\n", '<br>', "\n\n")

calls <- read.table(list.files(file.path(celltype), pattern="l1.csv", full.names=TRUE), sep=',', header=FALSE)
colnames(calls) <- c("Cell Type", "Cell Count")
calls %>% flextable() %>% autofit()

myimages<-list.files(file.path(celltype), pattern = "l1.png", full.names = TRUE)
#include_graphics(myimages[[2]])
#include_graphics(myimages[[1]])
include_graphics(myimages)

if(length(list.files(file.path(celltype), pattern = "l1_Split.png", full.names = TRUE)) > 0) {
  include_graphics(list.files(file.path(celltype), pattern = "l1_Split.png", full.names = TRUE)[[1]])
}

cat ("\n\n")
cat("### Level 2")
cat("\n\n")
calls <- read.table(list.files(file.path(celltype), pattern="l2.csv", full.names=TRUE), sep=',', header=FALSE)
colnames(calls) <- c("Cell Type", "Cell Count")
calls %>% flextable()# %>% autofit()

myimages<-list.files(file.path(celltype), pattern = "l2.png", full.names = TRUE)
include_graphics(myimages)

if(length(list.files(file.path(celltype), pattern = "l2_Split.png", full.names = TRUE)) > 0) {
  include_graphics(list.files(file.path(celltype), pattern = "l2_Split.png", full.names = TRUE)[[1]])
}

cat ("\n\n")
cat("### Level 3")
cat("\n\n")
calls <- read.table(list.files(file.path(celltype), pattern="l3.csv", full.names=TRUE), sep=',', header=FALSE)
colnames(calls) <- c("Cell Type", "Cell Count")
calls %>% flextable() %>% autofit()

myimages<-list.files(file.path(celltype), pattern = "l3.png", full.names = TRUE)
include_graphics(myimages)

if(length(list.files(file.path(celltype), pattern = "l3_Split.png", full.names = TRUE)) > 0) {
  include_graphics(list.files(file.path(celltype), pattern = "l3_Split.png", full.names = TRUE)[[1]])
}

```

<div class="tocify-extend-page" data-unique="tocify-extend-page" style="height: 0;"></div>
